Integrate CMSSW with HTCondor's update service.

bbockelm opened the pull request:
This commit provides a new default service, CondorStatusService,
which automatically reports basic progress statistics (# events,

The service will automatically detect if it's running as part of
a HTCondor job and probes to see if HTCondor has user-level updates
enabled (and is a sufficient version).  If either check fails, then
the service does not register any callbacks with the framework and
effectively becomes a no-op.

To see if it is within a HTCondor job, the service looks for the
_CONDOR_CHIRP_CONFIG environment variable - a very cheap check.
To see if HTCondor supports this feature, it spawns a new process
and looks at the exit code (more expensive).

This uses the 'set_job_attr_delayed' mechanism of HTCondor,
which causes these updates to 'tag along' with the existing updates
for memory / disk / CPU usage.  Hence, the extra cost is the few
extra bytes in the update packet that goes out once every 5 minutes.
The update only goes as far as the local daemon; condor_chirp does
not wait for it to propagate to the remote host.  Hence, it exits
rapidly.

CMSSW will only update once every updateIntervalSeconds (defaults
to 15 minutes).  The HTCondor worker node and central components
all have additional rate limiting mechanisms to prevent overload.

bbockelm committed:
Integrate CMSSW with HTCondor's update service.

This commit provides a new default service, CondorStatusService,
which automatically reports basic progress statistics (# events,

The service will automatically detect if it's running as part of
a HTCondor job and probes to see if HTCondor has user-level updates
enabled (and is a sufficient version).  If either check fails, then
the service does not register any callbacks with the framework and
effectively becomes a no-op.

To see if it is within a HTCondor job, the service looks for the
_CONDOR_CHIRP_CONFIG environment variable - a very cheap check.
To see if HTCondor supports this feature, it spawns a new process
and looks at the exit code (more expensive).

This uses the 'set_job_attr_delayed' mechanism of HTCondor,
which causes these updates to 'tag along' with the existing updates
for memory / disk / CPU usage.  Hence, the extra cost is the few
extra bytes in the update packet that goes out once every 5 minutes.
The update only goes as far as the local daemon; condor_chirp does
not wait for it to propagate to the remote host.  Hence, it exits
rapidly.

CMSSW will only update once every updateIntervalSeconds (defaults
to 15 minutes).  The HTCondor worker node and central components
all have additional rate limiting mechanisms to prevent overload.

cmsbuild commented:
A new Pull Request was created by @bbockelm (Brian Bockelman) for CMSSW_7_6_X.

Integrate CMSSW with HTCondor's update service.

It involves the following packages:

FWCore/Framework
FWCore/Services


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

bbockelm committed:
Extend attributes reported to HTCondor.

This commit adds a few more reported attributes:

- ChirpCMSSWFinished: Unix timestamp of when the job has finished.
- ChirpCMSSWLastUpdate: Unix timestamp of when the last update occurred.
- ChirpCMSSWMaxEvents: Maximum number of input events CMSSW is
  configured to process.  From the process's maxEvents pset.
- ChirpCMSSWMaxLumis: Maximum number of lumis CMSSW is configured
  to process.  From process's maxLuminosityBlocks pset and some
  simple processing of the source.

If no max is configured, the attribute is not reported.

The motivation behind these attributes is they:
- Simplify deadlock detection.  Using the defaults, the LastUpdate
  attribute should never be more than 30 minutes if Finished isn't set.
- Provide simple estimates of percent completion.  When we can determine
  the number of events or lumis to process (something we can for the
  majority of the grid use cases), we'll be able to determine the
  number of events/lumis left to process and an aggregate event/lumi
  processing rate.

bbockelm commented:
I added a few new attributes:

```
ChirpCMSSWFinished = 1436240094
ChirpCMSSWLastUpdate = 1436240094
ChirpCMSSWMaxEvents = 5
ChirpCMSSWMaxLumis = 5
```

These allow us to:
- Better detect deadlocks.  LastUpdate and Finished will allow us to detect whether CMSSW is running; if so, how long it has been since an event was completed.
- Provide rough estimates of percent completion.  The logic to detect MaxEvents / MaxLumis from the pset is far from perfect, but ought to work for 99% of the grid jobs out there (the core use case).

cmsbuild commented:
Pull request #10056 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Framework/bin/cmsRun.cpp:
```diff
@@ -332,6 +332,7 @@ int main(int argc, char* argv[]) {
       defaultServices.push_back("AdaptorConfig");
       defaultServices.push_back("SiteLocalConfigService");
       defaultServices.push_back("StatisticsSenderService");
+      defaultServices.push_back("CondorStatusService");
```
Why make this a default service? Why not require WMAgent to add it to the configurations since this is only needed on grid jobs?

Dr15Jones commented on FWCore/Services/src/CondorStatusUpdater.cc:
```diff
@@ -0,0 +1,272 @@
+void
+CondorStatusService::beginPost()
+{
+    ParameterSet const& processParameterSet = edm::getProcessParameterSet();
```
@wddgit is this the function you've been working to get rid of?

bbockelm replied:
I had several attempts to avoid using this function, but none were successful.

I would be quite happy if you had some guidance on how I could otherwise get access to the main source's PSet.

Dr15Jones commented:
Why not have CondorStatusUpdater.cc also hold the content of CondorStatusUpdater.h and move it to plugins and then register the service directly in  CondorStatusUpdater.cc? As far as I can tell there is no need for anyone to attempt to talk directly to  CondorStatusUpdater.

cmsbuild commented:
+1
Tested at: ddf76262fee9549a2262da7108a3f5c72e9f3e4e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/5957/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b8c2bd2d4b39b9c84f5a7e996eb31b0c4a80f705
https://github.com/cms-sw/cmssw/commit/16bc21a173968487d1022f106a048ecf69c5674b
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/5957/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/5957/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/5957/summary.html

bbockelm committed:
Move CondorStatusUpdater to plugins.

bbockelm committed:
Verify we overwrite all attributes at startup.

If a single HTCondor job has multiple CMSSW processes run sequentially
(i.e., GEN-SIM step followed by DIGI-RECO in the same job), then
attributes set at the end of one process will still be around when
the next one starts up.  Hence, we overwrite all attributes at
startup, even if we don't have meaningful values yet (we don't know
the maxEvents until BeginJob).

Notice we set ChirpCMSSWDone last; this is because the attribute
setting is ordered but not atomic.  Hence, we can write policies
and prefix them with:

(ChirpCMSSWDone=!=true) && ...

and know that any further attribute references will belong to the
current running job.

bbockelm commented:
@Dr15Jones - took care of the move to plugins.  I also updated this to be more resilient / useful in the case where multiple cmsRuns are invoked sequentially within the same HTCondor job.

I thought a lot about whether to enable this by default or have WMAgent enable it.  I chose the 'default' route because:
- The cost to detect the presence of HTCondor is extremely cheap (simple call to getenv).  There is no event-loop overhead if HTCondor is not present.
- While the core target is grid jobs, this is actually useful at any HTCondor site where the user might invoke cmsRun directly.  Hence, I view the scope as wider than WMAgent.

cmsbuild commented:
Pull request #10056 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

bbockelm committed:
Report number of files to process to HTCondor.

bbockelm committed:
Found misspelled attribute in testing.

wddgit commented:
Yes. I was trying to get rid of edm::getProcessParameterSet().
The current implementation of that function does not work
with SubProcessess's and requires a mutex for multithreaded
jobs. Here is what I replaced it with in PrescaleService:

void PrescaleService::preBeginJob(PathsAndConsumesOfModulesBase const&,
                                                        ProcessContext const& processContext) {
      if(!processParameterSetID_.isValid()) {
        processParameterSetID_ = processContext.parameterSetID();
      }
    }
 
    void PrescaleService::postBeginJob() {
      ParameterSet const& prcPS = edm::getParameterSet(processParameterSetID_);
...

cmsbuild commented:
Pull request #10056 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

bbockelm commented:
Yeah, I can adapt to that.  Give me a few minutes to update the PR.

wddgit commented:
Good.  The getParameterSet function is in ParameterSet.h.  The part in postBeginJob can actually go anywhere after preBeginJob or even in preBeginJob. Also the part about checking the validity of the ParameterSetID is important because preBeginJob can be called multiple times in a SubProcess
job and you would want the first one.

bbockelm committed:
Replace use of getProcessParameterSet in CondorStatusService.

cmsbuild commented:
Pull request #10056 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

bbockelm commented:
@wddgit can you review cdf7513 to make sure I did it right?  Tests OK, but I don't have a full range of process / sub-process complexity in my tests.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented:
I looked at the part related to getting the process parameter set and it looks good to me.

cmsbuild commented:
+1
Tested at: cdf7513e9b606d1d127995fdc09838e732ef43a5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/8538/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b8c2bd2d4b39b9c84f5a7e996eb31b0c4a80f705
https://github.com/cms-sw/cmssw/commit/16bc21a173968487d1022f106a048ecf69c5674b
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/8538/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/8538/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/8538/summary.html

bbockelm commented:
@Dr15Jones - github's history is a touch garbled.  Did I address all the items?

Dr15Jones commented:
@davidlange6 I'm not thrilled with having this be a default service since it's primary job is to monitor grid jobs (although Brian also thinks there are corner cases outside of that where it might be useful). The alternative would require WMAgent/CRAB to add the Service into the configuration. Where would you prefer this Service be turned on?

Dr15Jones commented:
@davidlange6 just a reminder that when we talked in person that you would comment to this thread

davidlange6 commented:
@bbockelm - @dr15jones - this would be much easier to put in 74x if its an optional service that we can first enable for a test. Being on by default makes that difficult. Of course it can be made a default service later once testing is done.

> On Jul 7, 2015, at 3:23 PM, Brian Bockelman <notifications@github.com> wrote:
> 
> @Dr15Jones - took care of the move to plugins. I also updated this to be more resilient / useful in the case where multiple cmsRuns are invoked sequentially within the same HTCondor job.
> 
> I thought a lot about whether to enable this by default or have WMAgent enable it. I chose the 'default' route because:
> 
> 	• The cost to detect the presence of HTCondor is extremely cheap (simple call to getenv). There is no event-loop overhead if HTCondor is not present.
> 	• While the core target is grid jobs, this is actually useful at any HTCondor site where the user might invoke cmsRun directly. Hence, I view the scope as wider than WMAgent.
> —
> Reply to this email directly or view it on GitHub.
>

hufnagel commented:
For testing alone you would not need to have WMAgent/CRAB support. You could just run a configuration that enables the service. Longer term this is of course not an option if you want to have it on for all central production and analysis and don't want to turn it on by default. But its enough for initial testing...

Dr15Jones commented:
@bbockelm once the setting of this on by default is changed, we'll be able to integrate it.

bbockelm committed:
Disable service by default pending testing.

bbockelm commented:
For potential backporting - should I just do a new PR with a squashed commit?

cmsbuild commented:
Pull request #10056 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
I recommended not squashing a back port since if we later do an auto forward port git will be confused.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
-1 
Tested at: acd0eea88793aac46b911cdf1ea521f9cb289c09
 When I ran the RelVals I found an error in the following worklfows: 
 50202.0 step2
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50+MINIAODMCUP1550/step2_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50+MINIAODMCUP1550.log

----- Begin Fatal Exception 21-Jul-2015 16:57:31 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Constructing the EventProcessor

   [1] Constructing module: class=MixingModule label='mix'

Exception Message:

RootEmbeddedFileSequence no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>
25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+MINIAODMCUP15/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+MINIAODMCUP15.log

----- Begin Fatal Exception 21-Jul-2015 17:15:11 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Constructing the EventProcessor

   [1] Constructing module: class=MixingModule label='mix'

Exception Message:

RootEmbeddedFileSequence no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10056/6402/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

bbockelm commented:
@Dr15Jones - how do you suggest I do the backport without squashing?  Does CMS have a how-to page for that (I always squash...)

Dr15Jones commented:
I've always just used ```git cherry-pick```. I think the new version of git even allows you to specify multiple commits in one command.

bbockelm commented:
Ah, ok.  I thought there was a deeper secret than that.
